### PR TITLE
[2.12] Configurable warnings (backported)

### PR DESCRIPTION
**Pull request by dwijnand**
Backport of:
- [x] https://github.com/scala/scala/pull/6635 prefer internal.Reporter
- [x] ~https://github.com/scala/scala/pull/6920 Harden DisplayReporter against null messages~ skipped, it's in "simplify reporters"
- [x] ~https://github.com/scala/scala/pull/6778 Revert to using ConsoleReporter during spin up~ should have cherry-picked, but I realised after I'd basically manually copied this over while resolving the conflicts in cherry-picking "simplify reporters"
- [x] https://github.com/scala/scala/pull/8338 simplify reporters
- [x] https://github.com/scala/scala/pull/8373 -Wconf
- [x] https://github.com/scala/scala/pull/8385 More source compatibility tweaks for reporters
- [x] https://github.com/scala/scala/pull/8852 Fix warning site for unused warnings
- [x] https://github.com/scala/scala/pull/8923 Warning summary respects -nowarn
- [x] https://github.com/scala/scala/commit/649fea5d13018b4d26b2677a881058b50dda63ad "Back out clearSetByUser"
- [x] https://github.com/scala/scala/pull/9032 Fix `@nowarn` to use correct semantics for `&`
- [x] ~https://github.com/scala/scala/pull/7756 REPL summarizes warnings~ nvm, deferred as it's more ergonomics than correctness
- [x] https://github.com/scala/scala/pull/7714 Promote -deprecation to -Xlint:deprecation

### CODE REVIEW

**Comment by dwijnand** [comment]
The -Wconf commit is failing over:

```
% diff /home/jenkins/workspace/scala-2.12.x-validate-main/test/files/run/iterator-from.check /home/jenkins/workspace/scala-2.12.x-validate-main/test/files/run/iterator-from-run.log--- empty
+++ iterator-from-run.log
@@ -1,0 +1,1 @@
+warning: two optimizer warnings; re-run enabling -opt-warnings for details, or try -help
	at scala.tools.partest.sbt.SBTRunner.makeStatus(SBTRunner.scala:87)
	at scala.tools.partest.sbt.SBTRunner$$anon$1.<init>(SBTRunner.scala:75)
	at scala.tools.partest.sbt.SBTRunner.onFinishTest(SBTRunner.scala:71)
	at scala.tools.partest.nest.AbstractRunner.runTest(AbstractRunner.scala:329)
	at scala.tools.partest.nest.AbstractRunner.$anonfun$runTestsForFiles$2(AbstractRunner.scala:336)
	at scala.tools.partest.package$$anon$2.call(package.scala:141)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

and

```
% diff /home/jenkins/workspace/scala-2.12.x-validate-main/test/files/run/mapConserve.check /home/jenkins/workspace/scala-2.12.x-validate-main/test/files/run/mapConserve-run.log--- empty
+++ mapConserve-run.log
@@ -1,0 +1,1 @@
+warning: one optimizer warning; re-run enabling -opt-warnings for details, or try -help
	at scala.tools.partest.sbt.SBTRunner.makeStatus(SBTRunner.scala:87)
	at scala.tools.partest.sbt.SBTRunner$$anon$1.<init>(SBTRunner.scala:75)
	at scala.tools.partest.sbt.SBTRunner.onFinishTest(SBTRunner.scala:71)
	at scala.tools.partest.nest.AbstractRunner.runTest(AbstractRunner.scala:329)
	at scala.tools.partest.nest.AbstractRunner.$anonfun$runTestsForFiles$2(AbstractRunner.scala:336)
	at scala.tools.partest.package$$anon$2.call(package.scala:141)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

File: src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala [hunk 1]
```diff
@@ -139,7 +139,7 @@ abstract class Inliner {
       val warnings = inline(request)
       for (warning <- warnings) {
         if (warning.emitWarning(compilerSettings))
-          backendReporting.inlinerWarning(request.callsite.callsitePosition, warning.toString)
+          backendReporting.optimizerWarning(request.callsite.callsitePosition, warning.toString, site = "") // FIXME
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala]:
```suggestion
          backendReporting.optimizerWarning(request.callsite.callsitePosition, warning.toString, site = backendUtils.optimizerWarningSiteString(request.callsite))
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -1841,7 +1843,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           else if (clazz != SerialVersionUIDAttr) restrictionWarning(cdef.pos, unit,
             """|subclassing Classfile does not
                |make your annotation visible at runtime.  If that is what
-               |you want, you must write the annotation class in Java.""".stripMargin)
+               |you want, you must write the annotation class in Java.""".stripMargin, WarningCategory.Other, clazz) // FIXME
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: 👍 
```suggestion
               |you want, you must write the annotation class in Java.""".stripMargin, WarningCategory.Other, clazz)
```

File: src/compiler/scala/tools/reflect/FormatInterpolator.scala [hunk 3]
```diff
@@ -128,7 +128,7 @@ abstract class FormatInterpolator {
               c.error(errPoint, msg("unsupported"))
               s0
             } else {
-              currentRun.reporting.deprecationWarning(errPoint, msg("deprecated"), "2.11.0")
+              currentRun.reporting.deprecationWarning(errPoint, msg("deprecated"), "2.11.0", site = "", origin = "") // FIXME
```
Comment (lrytz) [inline src/compiler/scala/tools/reflect/FormatInterpolator.scala]: 👍 

```suggestion
              currentRun.reporting.deprecationWarning(errPoint, msg("deprecated"), "2.11.0", site = "", origin = "")
```

File: src/reflect/scala/reflect/internal/Variances.scala [hunk 4]
```diff
@@ -85,7 +85,7 @@ trait Variances {
             // Unsound pre-2.11 behavior preserved under -Xsource:2.10
             if (settings.isScala211 || sym.isOverridingSymbol) Invariant
             else {
-              currentRun.reporting.deprecationWarning(sym.pos, "Construct depends on unsound variance analysis and will not compile in scala 2.11 and beyond", "2.11.0")
+              currentRun.reporting.deprecationWarning(sym.pos, "Construct depends on unsound variance analysis and will not compile in scala 2.11 and beyond", "2.11.0", site = "", origin = "") // FIXME
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/Variances.scala]: 👍 

```suggestion
              currentRun.reporting.deprecationWarning(sym.pos, "Construct depends on unsound variance analysis and will not compile in scala 2.11 and beyond", "2.11.0", site = "", origin = "")
```

File: test/files/pos/nowarnMacros/Test_2.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// scalac: -Xfatal-warnings
+
+import language.experimental.macros
+import scala.annotation.nowarn
+
+// FIXME this is meant to be a neg test
```
Comment (lrytz) [inline test/files/pos/nowarnMacros/Test_2.scala]: the warnings here don't show up due to https://github.com/scala/scala/pull/8995

the test was modified in 2.13.x to trigger a different warning, see https://github.com/scala/scala/commit/42094282694d3d61d6f0baa7800b59929345485b#diff-85c0d60e1ca722bbccc975ea082b002140f4c5e4be351b8be53ef64b9e1d3844

File: src/compiler/scala/tools/nsc/Global.scala [hunk 6]
```diff
@@ -1363,14 +1363,14 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     private def warnDeprecatedAndConflictingSettings() {
       // issue warnings for any usage of deprecated settings
       settings.userSetSettings filter (_.isDeprecated) foreach { s =>
-        currentRun.reporting.deprecationWarning(NoPosition, s.name + " is deprecated: " + s.deprecationMessage.get, "")
+        runReporting.deprecationWarning(NoPosition, s.name + " is deprecated: " + s.deprecationMessage.get, "", "", "")
       }
       val supportedTarget = "jvm-1.8"
       if (settings.target.value != supportedTarget) {
-        currentRun.reporting.deprecationWarning(NoPosition, settings.target.name + ":" + settings.target.value + " is deprecated and has no effect, setting to " + supportedTarget, "2.12.0")
+        runReporting.deprecationWarning(NoPosition, settings.target.name + ":" + settings.target.value + " is deprecated and has no effect, setting to " + supportedTarget, "2.12.0", site = "", origin = "") // FIXME
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/Global.scala]: 👍 

```suggestion
        runReporting.deprecationWarning(NoPosition, settings.target.name + ":" + settings.target.value + " is deprecated and has no effect, setting to " + supportedTarget, "2.12.0", site = "", origin = "")
```

**Review by lrytz — COMMENTED** [review]
Looked through the `FIXME`s. I pushed a change for the ones in `WConfTest`, commented on the others.

**Review by lrytz — APPROVED** [review]
LGTM otherwise - I'm impressed how you pushed this through! I really didn't think this was feasible so cleanly and in such a short time, you proved me wrong 🙂

**Comment by SethTisue** [comment]
needs community build work: https://github.com/scala/community-build/pull/1265

File: test/files/run/t1931.check [hunk 7]
```diff
@@ -10,8 +10,8 @@ import Predef.{any2stringadd=>_, _}
 
 scala> x + " works"
 <console>:14: error: value + is not a member of Any
-       x + " works"
-         ^
+x + " works"
+  ^
```
Comment (som-snytt) [inline test/files/run/t1931.check:14]: Noticed that this is a regression. The text is supposed to align with text at prompt. I haven't checked the current head.

**Pull request merged**